sdk/helper/nonce/encrypted_nonce.go

```diff
@@ -0,0 +1,374 @@
+		return "", fmt.Errorf("failed to read AEAD nonce: %w", err)
+	}
+
+	plaintext := make([]byte, noncePlaintextLength)
+	binary.BigEndian.PutUint64(plaintext[0:], counter)
+	binary.BigEndian.PutUint64(plaintext[8:], uint64(expiry.Unix()))
+	ciphertext := ens.crypt.Seal(nil, nonce, plaintext, nil)
+
+	// Now, generate the wire format of the nonce. Use a prefix, the nonce,
+	// and then the ciphertext.
+	var wire []byte
+	wire = append(wire, []byte(nonceSentinel)...)
+	wire = append(wire, nonce[4:]...)
+	wire = append(wire, ciphertext...)
+
+	if len(wire) != nonceLength {
+		return "", fmt.Errorf("expected nonce length of %v got %v", nonceLength, len(wire))
+	}
+
+	return base64.RawURLEncoding.EncodeToString(wire), nil
+}
+
+func (ens *encryptedNonceService) recordCounterForTime(counter uint64, expiry time.Time) {
+	timestamp := ensTimestamp(expiry.Unix())
+	value := ensCounter(counter)
+
+	ens.issueLock.Lock()
+	defer ens.issueLock.Unlock()
+
+	lastValue, ok := ens.maxIssued[timestamp]
+	if !ok || lastValue < value {
+		ens.maxIssued[timestamp] = value
+	}
+}
+
+func (ens *encryptedNonceService) Get() (token string, expiry time.Time, err error) {
+	counter := ens.nextCounter.Add(1)
+	now := time.Now()
+	then := now.Add(ens.validity)
+
+	token, err = ens.encryptNonce(counter, then)
+	if err != nil {
+		return "", now, err
+	}
+
+	ens.recordCounterForTime(counter, then)
+	return token, then, nil
+}
+
+func (ens *encryptedNonceService) decryptNonce(token string) (counter uint64, expiry time.Time, ok bool) {
+	zero := time.Time{}
+
+	wire, err := base64.RawURLEncoding.DecodeString(token)
+	if err != nil {
+		return 0, zero, false
+	}
+
+	if len(wire) != nonceLength {
+		return 0, zero, false
+	}
+
+	data := wire
+
+	sentinel := data[0:len(nonceSentinel)]
+	data = data[len(nonceSentinel):]
+	if subtle.ConstantTimeCompare([]byte(nonceSentinel), sentinel) != 1 {
+		return 0, zero, false
+	}
+
+	nonce := make([]byte, 12)
+	for i := 0; i < 4; i++ {
+		nonce[i] = 0
+	}
+	copy(nonce[4:12], data[0:8])
+	data = data[8:]
+
+	ciphertext := data[:]
+
+	plaintext, err := ens.crypt.Open(nil, nonce, ciphertext, nil)
+	if err != nil {
+		return 0, zero, false
+	}
+
+	if len(plaintext) != noncePlaintextLength {
+		return 0, zero, false
+	}
+
+	counter = binary.BigEndian.Uint64(plaintext[0:8])
+	unix := binary.BigEndian.Uint64(plaintext[8:])
+	expiry = time.Unix(int64(unix), 0)
+
+	return counter, expiry, true
+}
+
+func (ens *encryptedNonceService) Redeem(token string) bool {
+	now := time.Now()
+	counter, expiry, ok := ens.decryptNonce(token)
+	if !ok {
+		return false
+	}
+
+	if expiry.Before(now) {
+		return false
+	}
+
+	if counter <= ens.minCounter.Load() {
+		return false
+	}
+
+	timestamp := ensTimestamp(expiry.Unix())
+	counterValue := ensCounter(counter)
+
+	// From here on out, we're doing the expensive checks. This _looks_
+	// like a valid token, but now we want to verify the used-exactly-once
+	// nature.
```

Comment:
Does this work in multiple nodes in a cluster?

Reply:
This one doesn't, because it sets `IsMultiCluster() = false`. Vault ACME is only on the perf active node and isn't cross-perf-cluster.

Reply:
Even in a failover of the primary to another node?  Presumably that'd just allow for the rare double accept.

Reply:
Each node will have separate AES keys because Initialize generates random keys and there's no external key provisioning API, so in the nonces from another node will not be decryptable by this node. This prevents double-accept I believe. The downside of course, is that nonces wouldn't be redeemable from before the failover, but ACME clients already re-fetch nonces in the event of a bad nonce.

For other usages, if this is desirable, we could perhaps have a stored-key version of this, that has the double-accept problem but uses a long-term stored key (which does introduce other problems around key rotation fwiw...).